packages/fxa-settings/src/lib/url.ts

```diff
@@ -0,0 +1,180 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// A set of utilities to deal with URLs
+
+/**
+ * Convert a search string to its object representation, one entry
+ * per query parameter. Assumes the string is a search string and
+ * not a full URL without a search string.
+ */
+export function searchParams(str = '', allowedFields?: string[]) {
+  // ditch everything before the ? and from # to the end
+  const search = str.replace(/(^.*\?|#.*$)/g, '').trim();
+  if (!search) {
+    return {};
+  }
+
+  return splitEncodedParams(search, allowedFields!);
+}
+
+/**
+ * Return the value of a single query parameter in the string
+ */
+export function searchParam(name: string, str?: string): string {
+  return searchParams(str)[name];
+}
+
+/**
+ * Convert a hash string to its object representation, one entry
+ * per query parameter
+ */
+export function hashParams(str = '', allowedFields?: string[]) {
+  // ditch everything before the #
+  const hash = str.replace(/^.*#/, '').trim();
+  if (!hash) {
+    return {};
+  }
+
+  return splitEncodedParams(hash, allowedFields);
+}
+
+/**
+ * Convert a URI encoded string to its object representation.
+ *
+ * `&` is the expected delimiter between parameters.
+ * `=` is the delimiter between a key and a value.
+ */
+export function splitEncodedParams(str = '', allowedFields?: string[]) {
+  const pairs = str.split('&');
+  const terms: { [key: string]: string } = {};
+
+  pairs.forEach((pair) => {
+    const [key, value] = pair.split('=');
+    terms[key] = decodeURIComponent(value).trim();
+  });
+
+  if (!allowedFields) {
+    return terms;
+  }
+
+  return Object.keys(terms)
+    .filter((key) => allowedFields.indexOf(key) >= 0)
+    .reduce((newObj, key) => Object.assign(newObj, { [key]: terms[key] }), {});
+}
+
+/**
+ * Convert an object to a search string.
+ */
+export function objToSearchString(obj: {}) {
+  return objToUrlString(obj, '?');
+}
+
+/**
+ * Convert an object to a hash string.
+ */
+export function objToHashString(obj: {}) {
+  return objToUrlString(obj, '#');
+}
+
+/**
+ * Convert an object to a URL safe string
+ */
+export function objToUrlString(obj: { [key: string]: any } = {}, prefix = '?') {
+  const params: string[] = [];
+
+  for (const paramName in obj) {
+    if (paramName) {
+      const paramValue = obj[paramName];
+      if (paramValue != null && paramValue.length) {
+        params.push(paramName + '=' + encodeURIComponent(paramValue));
+      }
+    }
+  }
+
+  if (!params.length) {
+    return '';
+  }
+
+  return prefix + params.join('&');
+}
+
+/**
+ * Get the origin portion of the URL
+ */
+export function getOrigin(url: string): string | null {
+  if (!url) {
+    return '';
+  }
+
+  // The URL API is only supported by new browsers, a workaround is used.
+  const anchor = document.createElement('a');
```

Comment:
It might be worth another pass through some of these tests to see if any have portions we can rip out, like if we need this or not? And do we support Firefox 18 at this point (line 114)? 🤔 Do we need all of these if we have `url.js` in `fxa-content-server` when this is being served on the same domain, and are we even using this in `fxa-settings`? When I search for some of these functions I don't see them being called. Maybe we do need all of these, just asking questions and throwing things out there.

Reply:
> When I search for some of these functions I don't see them being called.

You are correct - some of them are not used. Given that it's written as a whole lib I didn't want to pull certain bits out because we weren't using them right now. But you're right, a lot of this isn't being used. I'm happy to pull everything out we're not using; we may need to just add it down the line.

Reply:
> are we even using this in fxa-settings?

`storage.ts` uses it. Only one part of it.